src/product.service.ts

```diff
@@ -0,0 +1,41 @@
+import { Product } from './product.types';
+
+export class ProductService {
+    private products: Product[] = [];
+
+    public addProduct(product: Product): void {
+        this.products.push(product);
+    }
+
+    public find(name: string): Product | undefined {
+        for (let i = 0; i < this.products.length; i++) {
+            if (this.products[i].name.toLowerCase() === name.toLowerCase()) {
+                return this.products[i];
+            }
+        }
+        return undefined;
+    }
+
+    public filter(category: string): Product[] {
+        const filteredProducts: Product[] = [];
+        for (let i = 0; i < this.products.length; i++) {
+            if (this.products[i].category === category) {
+                filteredProducts.push(this.products[i]);
+            }
+        }
+        return filteredProducts;
+    }
+
+    public search(tag: string): Product[] {
+        const result: Product[] = [];
+        for (let i = 0; i < this.products.length; i++) {
+            for (let j = 0; j < this.products[i].tags.length; j++) {
+                if (this.products[i].tags[j].toLowerCase() === tag.toLowerCase()) {
```

Comment:
As above tag.toLowerCase() should be done and stored in variable before tags for-loop to save runtime.

As improvement:
We could think about storing tags only as lowerCase strings; this would allow use to use this.products[i].includes(tag.toLowerCase()) which is much cleaner than the inner for-loop